Balancer-Bots/Large-Internal-Balance-Deposit-Withdrawal/src/agent.ts

```diff
@@ -0,0 +1,65 @@
+import { Finding, HandleBlock, BlockEvent, getEthersProvider } from "forta-agent";
+import { providers, BigNumber, utils } from "ethers";
+import BalanceFetcher from "./balance.fetcher";
+import { NetworkManager } from "forta-agent-tools";
+import { createFinding, NetworkData } from "./utils";
+import { EVENT } from "./constants";
+import CONFIG from "./agent.config";
+
+const networkManager = new NetworkManager<NetworkData>(CONFIG);
+
+export const initialize = (networkManager: NetworkManager<NetworkData>, provider: providers.Provider) => {
+  return async () => {
+    await networkManager.init(provider);
+  };
+};
+
+export const provideHandleBlock = (
+  provider: providers.Provider,
+  networkManager: NetworkManager<NetworkData>,
+  balanceFetcher: BalanceFetcher
+): HandleBlock => {
+  const vaultIface = new utils.Interface(EVENT);
+
+  const sighash = [vaultIface.getEventTopic("InternalBalanceChanged")];
+
+  return async (blockEvent: BlockEvent): Promise<Finding[]> => {
+    const findings: Finding[] = [];
+
+    const logs = (
+      await provider.getLogs({
+        address: networkManager.get("vaultAddress"),
+        fromBlock: blockEvent.blockNumber,
+        toBlock: blockEvent.blockNumber,
+      })
+    ).filter((log) => sighash.includes(log.topics[0]));
+
+    await Promise.all(
+      logs.map(async (log) => {
+        const decodedLog = vaultIface.parseLog(log);
+
+        const delta: BigNumber = BigNumber.from(decodedLog.args.delta);
+        balanceFetcher.setData(decodedLog.args.token);
+
+        // fetch token balance of the contract then set threshold.
+        const totalBalance: BigNumber = await balanceFetcher.getBalance(
+          blockEvent.blockNumber - 1,
+          networkManager.get("vaultAddress")
+        );
+
+        const _threshold = BigNumber.from(totalBalance).mul(networkManager.get("threshold")).div(100);
+
+        if (delta.abs().gte(_threshold)) {
+          findings.push(createFinding(decodedLog.args));
```

Comment:
```suggestion
  const topics = [vaultIface.getEventTopic("InternalBalanceChanged")];

  return async (blockEvent: BlockEvent): Promise<Finding[]> => {
    const findings: Finding[] = [];

    const logs = (
      await provider.getLogs({
        address: networkManager.get("vaultAddress"),
        topics, 
        fromBlock: blockEvent.blockNumber,
        toBlock: blockEvent.blockNumber,
      })
    ).map((log) => vaultIface.parseLog(log));

    await Promise.all(
      logs.map(async (log) => {
        const delta: BigNumber = BigNumber.from(log.args.delta);
        balanceFetcher.setData(log.args.token);

        // fetch token balance of the contract then set threshold.
        const totalBalance: BigNumber = await balanceFetcher.getBalance(
          blockEvent.blockNumber - 1,
          networkManager.get("vaultAddress")
        );

        const threshold = BigNumber.from(totalBalance).mul(networkManager.get("threshold")).div(100);

        if (delta.abs().gte(threshold)) {
          findings.push(createFinding(log.args));
```

Reply:
Also, it would be nice if `bignumber.js` `BigNumber` is used. The changes would be minimal and the config threshold field would support decimals.

Reply:
thanks for adding topic and shortening the logic. 

about bignumber.js, is it because the calculation is done more precisely?

Reply:
Yes but not exactly that, it's because `bignumber.js` supports decimals, so it wouldn't be a problem if the config threshold is set to one.